spatialos-sdk/src/worker/entity_builder.rs

```diff
@@ -0,0 +1,161 @@
+use crate::worker::{
+    component::Component,
+    component::ComponentId,
+    entity::Entity,
+    internal::schema::{
+        SchemaComponentData, SchemaDouble, SchemaObject, SchemaObjectField, SchemaPrimitiveField,
+        SchemaString, SchemaStringField, SchemaUint32,
+    },
+};
+use std::collections::HashMap;
+
+const ENTITY_ACL_COMPONENT_ID: ComponentId = 50;
+const METADATA_COMPONENT_ID: ComponentId = 53;
+const POSITION_COMPONENT_ID: ComponentId = 54;
+const PERSISTENCE_COMPONENT_ID: ComponentId = 55;
+
+pub struct EntityBuilder {
+    entity: Entity,
+
+    position: (f64, f64, f64),
+    is_persistent: bool,
+    metadata: Option<String>,
+
+    write_permissions: HashMap<ComponentId, String>,
+    read_permissions: Vec<String>,
+
+    error: Option<String>,
+}
+
+impl EntityBuilder {
+    pub fn new<T: Into<String>>(x: f64, y: f64, z: f64, position_write_layer: T) -> Self {
+        let builder = EntityBuilder {
+            entity: Entity::new(),
+            is_persistent: false,
+            metadata: None,
+            position: (x, y, z),
+            write_permissions: HashMap::new(),
+            read_permissions: Vec::new(),
+            error: None,
+        };
+
+        builder.set_write_access(POSITION_COMPONENT_ID, position_write_layer)
+    }
+
+    pub fn add_component<C: Component, T: Into<String>>(mut self, data: C, write_layer: T) -> Self {
+        if let Err(e) = self.entity.add(data) {
+            self.error = Some(e);
+        };
+
+        self.write_permissions.insert(C::ID, write_layer.into());
+        self
+    }
+
+    pub fn set_persistent<T: Into<String>>(mut self, write_layer: T) -> Self {
+        self.is_persistent = true;
+        self.set_write_access(PERSISTENCE_COMPONENT_ID, write_layer)
+    }
+
+    pub fn set_metadata<T: Into<String>, U: Into<String>>(
+        mut self,
+        entity_type: T,
+        write_layer: U,
+    ) -> Self {
+        self.metadata = Some(entity_type.into());
+        self.set_write_access(METADATA_COMPONENT_ID, write_layer)
+    }
+
+    pub fn set_read_access<T: AsRef<str>>(mut self, layers: &[T]) -> Self {
```

Comment:
It would probably be better to change the generic bound here to `T: Into<String>`, since the only thing we ever do with the layer names here is convert them into `String`s.